src/query/expression/src/evaluator.rs

```diff
@@ -507,20 +507,50 @@ pub struct ConstantFolder<'a, Index: ColumnIndex> {
 }
 
 impl<'a, Index: ColumnIndex> ConstantFolder<'a, Index> {
-    pub fn new(
+    /// Fold a single expression, returning the new expression and the domain of the new expression.
+    pub fn fold(
+        expr: &Expr<Index>,
+        func_ctx: FunctionContext,
+        fn_registry: &'a FunctionRegistry,
+    ) -> (Expr<Index>, Option<Domain>) {
+        let input_domains = expr
+            .column_refs()
+            .into_iter()
+            .map(|(name, ty)| {
+                let domain = Domain::full(&ty);
+                (name, domain)
+            })
+            .collect();
+
+        let folder = ConstantFolder {
+            input_domains,
+            func_ctx,
+            fn_registry,
+        };
+
+        folder.fold_to_stable(expr)
+    }
+
+    /// Fold a single expression with columns' domain, and then return the new expression and the
+    /// domain of the new expression.
+    pub fn fold_with_domain(
```

Comment:
I also thought about this. Maybe we can use `Option` here and unwrap to full domain.

Reply:
Yes but I think  the `None` may seem confusing in `ConstantFolder::fold(&expr, None, func_ctx, &BUILTIN_FUNCTIONS)`